File: backend/app/models/moment.py
"""Moment domain model."""

from __future__ import annotations

from datetime import UTC, datetime, timedelta
from enum import StrEnum
from typing import Any, ClassVar

from beanie import Document
from pydantic import BaseModel, Field
from pymongo import ASCENDING, DESCENDING, TEXT, IndexModel


class MomentVisibility(StrEnum):
    PUBLIC = "public"
    PRIVATE = "private"
    UNLISTED = "unlisted"


class MomentStatus(StrEnum):
    PUBLISHED = "published"
    DRAFT = "draft"
    ARCHIVED = "archived"


class MomentAttachmentType(StrEnum):
    IMAGE = "image"
    LINK = "link"
    BOOK = "book"
    QUOTE = "quote"


class MomentAttachment(BaseModel):
    type: MomentAttachmentType = MomentAttachmentType.IMAGE
    url: str
    thumbnail_url: str | None = None
    title: str | None = None
    description: str | None = None
    meta: dict[str, Any] | None = None


class MomentLocation(BaseModel):
    name: str | None = None
    latitude: float | None = None
    longitude: float | None = None


class Moment(Document):
    """Short-form note document model."""

    user_id: int
    content: str = Field(..., min_length=1, max_length=2000)
    summary: str | None = None
    visibility: MomentVisibility = MomentVisibility.PUBLIC
    status: MomentStatus = MomentStatus.PUBLISHED
    mood: str | None = None
    tags: list[str] = Field(default_factory=list)
    attachments: list[MomentAttachment] = Field(default_factory=list)
    location: MomentLocation | None = None
    source: str | None = None
    is_pinned: bool = False
    allow_comment: bool = True
    like_count: int = 0
    comment_count: int = 0
    view_count: int = 0
    published_at: datetime | None = None
    created_at: datetime = Field(default_factory=lambda: datetime.now(UTC))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(UTC))
    deleted_at: datetime | None = None

    class Settings:
        name = "moments"
        use_cache = True
        cache_expiration_time = timedelta(seconds=10)
        indexes: ClassVar[list] = [
            IndexModel(
                [
                    ("visibility", ASCENDING),
                    ("status", ASCENDING),
                    ("deleted_at", ASCENDING),
                    ("is_pinned", DESCENDING),
                    ("published_at", DESCENDING),
                ],
                name="public_moment_feed_idx",
            ),
            IndexModel(
                [
                    ("user_id", ASCENDING),
                    ("status", ASCENDING),
                    ("updated_at", DESCENDING),
                ],
                name="user_moment_admin_idx",
            ),
            IndexModel(
                [("tags", ASCENDING), ("published_at", DESCENDING)],
                name="moment_tags_idx",
            ),
            IndexModel(
                [("content", TEXT), ("tags", TEXT)],
                name="moment_text_idx",
            ),
        ]

File: go-backend/internal/mongo/document/moment.go
package document

import "time"

type MomentVisibility string

const (
	MomentPublic   MomentVisibility = "PUBLIC"
	MomentPrivate  MomentVisibility = "PRIVATE"
	MomentUnlisted MomentVisibility = "UNLISTED"
)

type MomentStatus string

const (
	MomentPublished MomentStatus = "PUBLISHED"
	MomentDraft     MomentStatus = "DRAFT"
	MomentArchived  MomentStatus = "ARCHIVED"
)

type MomentAttachmentType string

const (
	MomentAttachmentImage MomentAttachmentType = "IMAGE"
	MomentAttachmentLink  MomentAttachmentType = "LINK"
	MomentAttachmentBook  MomentAttachmentType = "BOOK"
	MomentAttachmentQuote MomentAttachmentType = "QUOTE"
)

type MomentAttachment struct {
	Type         MomentAttachmentType `bson:"type"`
	URL          string               `bson:"url"`
	ThumbnailURL *string              `bson:"thumbnail_url"`
	Title        *string              `bson:"title"`
	Description  *string              `bson:"description"`
	Meta         map[string]any       `bson:"meta"`
}

type MomentLocation struct {
	Name      *string  `bson:"name"`
	Latitude  *float64 `bson:"latitude"`
	Longitude *float64 `bson:"longitude"`
}

type Moment struct {
	ID           string             `bson:"_id,omitempty"`
	UserID       int                `bson:"user_id"`
	Content      string             `bson:"content"`
	Summary      *string            `bson:"summary"`
	Visibility   MomentVisibility   `bson:"visibility"`
	Status       MomentStatus       `bson:"status"`
	Mood         *string            `bson:"mood"`
	Tags         []string           `bson:"tags"`
	Attachments  []MomentAttachment `bson:"attachments"`
	Location     *MomentLocation    `bson:"location"`
	Source       *string            `bson:"source"`
	IsPinned     bool               `bson:"is_pinned"`
	AllowComment bool               `bson:"allow_comment"`
	LikeCount    int                `bson:"like_count"`
	CommentCount int                `bson:"comment_count"`
	ViewCount    int                `bson:"view_count"`
	PublishedAt  *time.Time         `bson:"published_at"`
	CreatedAt    time.Time          `bson:"created_at"`
	UpdatedAt    time.Time          `bson:"updated_at"`
	DeletedAt    *time.Time         `bson:"deleted_at"`
}
